source/common/protobuf/utility.cc

```diff
@@ -505,6 +506,77 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+// Recursive helper method for MessageUtil::redact() below.
+void redactInPlace(Protobuf::Message* message) {
```

Comment:
drive by: this recursive logic is almost identical to what we need to do for deprecation checks, etc. Is there some generic recursion code we can write which would then apply a visitor pattern to actually do an action on a message/node?

Reply:
I think so, yes... I don't know if the traversal would work for `VersionConverter::upgrade` since that has to traverse two messages in parallel, but it should work for other cases. That said, I think we should do that sort of refactoring in a separate PR after this lands, if that's ok.

Reply:
Yes, let's do this in a followup once this lands and the v3 related PRs are in, since they are all doing something pretty similar.